src/com/amazon/corretto/crypto/provider/AesCbcSpi.java

```diff
@@ -0,0 +1,704 @@
+      return (int) (all - rem);
+    }
+    // When all data (inputLen + unprocessedInput) is block-size aligned, padding is enabled, and we
+    // are decrypting, the cipher does not decrypt the last block until doFinal.
+    return (int) (all - BLOCK_SIZE_IN_BYTES);
+  }
+
+  private int update(
+      final ByteBuffer inputDirect,
+      final byte[] inputArray,
+      final int inputOffset,
+      final int inputLen,
+      final ByteBuffer outputDirect,
+      final byte[] outputArray,
+      final int outputOffset) {
+
+    // Unlike, doFinal (which needs to decide if a context should be released or not), update always
+    // has to save the context.
+
+    final long[] ctxContainer = new long[] {0};
+    try {
+      final int result;
+      if (cipherState == CipherState.INITIALIZED) {
+        if (nativeCtx != null) {
+          result =
+              nativeCtx.use(
+                  ctxPtr ->
+                      nInitUpdate(
+                          opMode,
+                          padding,
+                          key,
+                          key.length,
+                          iv,
+                          null,
+                          ctxPtr,
+                          inputDirect,
+                          inputArray,
+                          inputOffset,
+                          inputLen,
+                          outputDirect,
+                          outputArray,
+                          outputOffset));
+        } else {
+          result =
+              nInitUpdate(
+                  opMode,
+                  padding,
+                  key,
+                  key.length,
+                  iv,
+                  ctxContainer,
+                  0,
+                  inputDirect,
+                  inputArray,
+                  inputOffset,
+                  inputLen,
+                  outputDirect,
+                  outputArray,
+                  outputOffset);
+          nativeCtx = new NativeEvpCipherCtx(ctxContainer[0]);
+        }
+        cipherState = CipherState.UPDATED;
+      } else {
+        // Cipher is in UPDATED state: this is not the first time update is being invoked.
+        result =
+            nativeCtx.use(
+                ctxPtr ->
+                    nUpdate(
+                        ctxPtr,
+                        inputDirect,
+                        inputArray,
+                        inputOffset,
+                        inputLen,
+                        unprocessedInput,
+                        outputDirect,
+                        outputArray,
+                        outputOffset));
+        // No need to update the cipherState since it's already in UPDATED state.
+      }
+      final long all = inputLen + unprocessedInput;
+      unprocessedInput = (int) (all - result);
+      return result;
+    } catch (final Exception e) {
+      cipherState = CipherState.NEEDS_INITIALIZATION;
+      saveNativeContextIfNeeded(ctxContainer[0]);
```

Comment:
if we're resetting the cipher state, why save the native context? shouldn't we reset all state?

Reply:
The more I think about it, doing the JNI call and cleaning it up is better. By registering it with Janitor, we will only save IF the same Cipher object is re used.